lib/Target/ExportSMTLIB/ExportSMTLIB.cpp

```diff
@@ -334,10 +339,28 @@ struct ExpressionVisitor
     newInfo.stream.indent(newInfo.indentLevel);
     if (failed(printExpression(worklist, newInfo)))
       return failure();
+
+    // llvm::outs()<<"\n\n\nEXPRESSION\n ";
+
+
+    // printExpression(worklist, newInfo);
+
     info.stream << info.valueMap.lookup(yieldedValue);
-    for (unsigned k = 0; k < newInfo.openParens + 1; ++k)
+
+    // llvm::outs()<<"\n\n\nOpen Parens is "<<newInfo.openParens;
+
+    if(weight != 0 && newInfo.openParens > 0)
+      info.stream <<") :weight "<< weight << ")";
+    else if (weight != 0)
+      info.stream <<" :weight "<< weight << "))";
+    else
+      info.stream << ")";
+
+    for (unsigned k = 0; k < newInfo.openParens; ++k)
```

Comment:
There might be a bug in my code above (the initialization of `newInfo`) where we inherit the number of open parens from outside the quantifier, but here we don't subtract that. Maybe we should initialize `openParens` for `newInfo` to 0. But this probably only affects your change if the number of let expressions outside the quantifier is more than 1.
(not expecting you to fix my bugs; just mentioning it such that you are aware of it)

Reply:
I'm not sure I understood what you mean, why should we subtract the number of parens from outside?

Reply:
Such that we don't close `let` scopes from outside the quantifier too early. The `assert` should close them later on instead.
You can try this example to trigger the issue:
```
  %2 = smt.exists {
  ^bb0(%arg2: !smt.int, %arg3: !smt.int):
    %3 = smt.eq %arg2, %arg3 : !smt.int
    smt.yield %3 : !smt.bool
  }
  %3 = smt.exists {
  ^bb0(%arg2: !smt.int, %arg3: !smt.int):
    %5 = smt.eq %arg2, %arg3 : !smt.int
    smt.yield %5 : !smt.bool
  }
  %5 = smt.and %2, %3
  smt.assert %5
```

Reply:
Clear, thanks a lot!